File: lecture5/solutions/daily_temp.java
import java.util.Stack;

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Integer> stack = new Stack<>();
        int[] ans = new int[temperatures.length];

        for (int i = 0; i < temperatures.length; i++) {
            while (!stack.isEmpty() && temperatures[stack.peek()] < temperatures[i]) {
                int index = stack.pop();
                ans[index] = i - index;
            }
            stack.push(i);
        }

        return ans;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] temperatures = {73, 74, 75, 71, 69, 72, 76, 73};
        int[] result = solution.dailyTemperatures(temperatures);

        for (int r : result) {
            System.out.print(r + " ");
        }
        // Output: 1 1 4 2 1 1 0 0
    }
}

File: lecture5/solutions/daily_temp.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        stack<int> st;
        vector<int> ans(temperatures.size(), 0);

        for (int i = 0; i < temperatures.size(); i++) {
            if (!st.empty()) {
                while (!st.empty() && temperatures[st.top()] < temperatures[i]) {
                    ans[st.top()] = i;
                    st.pop();
                }
            }

            st.push(i);
        }

        for (int i = 0; i < ans.size(); i++) {
            if (ans[i] > 0) {
                ans[i] = ans[i] - i;
            }
        }

        return ans;
    }
};